ntb_hw_switchtec.c

```diff
@@ -992,20 +992,84 @@ static int config_rsvd_lut_win(struct switchtec_ntb *sndev,
 	return 0;
 }
 
-static int config_req_id_table(struct switchtec_ntb *sndev,
-			       struct ntb_ctrl_regs __iomem *mmio_ctrl,
-			       int *req_ids, int count)
+static int add_req_id(struct switchtec_ntb *sndev,
+		      struct ntb_ctrl_regs __iomem *mmio_ctrl, int req_id)
 {
 	int i, rc = 0;
+	int slot = -1;
 	u32 error;
-	u32 proxy_id;
+	int table_size;
+	u32 proxy_id = 0;
+	bool exist = false;
+	bool added = true;
+
+	table_size = ioread16(&mmio_ctrl->req_id_table_size);
+
+	rc = switchtec_ntb_part_op(sndev, mmio_ctrl,
+				   NTB_CTRL_PART_OP_LOCK,
+				   NTB_CTRL_PART_STATUS_LOCKED);
+	if (rc)
+		return rc;
+
+	iowrite32(NTB_PART_CTRL_ID_PROT_DIS, &mmio_ctrl->partition_ctrl);
+
+	for (i = 0; i < table_size; i++) {
+		proxy_id = ioread32(&mmio_ctrl->req_id_table[i]);
 
-	if (ioread16(&mmio_ctrl->req_id_table_size) < count) {
+		if (!(proxy_id & NTB_CTRL_REQ_ID_EN) && (slot == -1))
+			slot = i;
+
+		if ((proxy_id & NTB_CTRL_REQ_ID_EN) &&
+		    (proxy_id >> 16) == req_id) {
+			exist = true;
```

Comment:
Maybe just goto an unlock tag, instead of adding the extra variable, condition and indentation.